File: shoring/fxapp/src/main/java/com/panopset/fxapp/FontManagerFX.java
package com.panopset.fxapp;

import java.util.ArrayList;
import java.util.List;
import com.panopset.compat.GlobalProperties;
import com.panopset.compat.GlobalPropertyKeys;
import com.panopset.compat.Logop;
import com.panopset.compat.Logop.LogEntry;
import javafx.application.Platform;
import javafx.scene.Node;
import javafx.scene.control.MenuBar;
import javafx.scene.control.Tab;
import javafx.scene.control.TabPane;
import javafx.scene.control.TextField;
import javafx.scene.control.TextInputControl;
import javafx.scene.text.Font;
import javafx.scene.text.FontPosture;
import javafx.scene.text.FontWeight;

public enum FontManagerFX implements GlobalPropertyKeys, FontNames {

  INSTANCE;

  public final static String FONT_GREEN = "-fx-background-color: #000000; -fx-text-fill: #99ff99;";
  public final static String FONT_YELLOW = "-fx-background-color: #000000;-fx-text-fill: #ffff66;";
  public final static String FONT_ORANGE = "-fx-background-color: #000000;-fx-text-fill: #ffaa00;";
  public final static String FONT_BLUE = "-fx-background-color: #000000;-fx-text-fill: #0000ff;";
  public final static String FONT_RED = "-fx-background-color: #000000;-fx-text-fill: #ff3333;";
  public final static String FONT_PURPLE = "-fx-background-color: #000000;-fx-text-fill: #dd33dd;";

  private Font monospace;
  private Font boldArial;
  private Font plainArial;
  private Font boldSerif;
  private Font plainSerif;
  private Font borderTitle;

  private LogEntry logEntry;

  public static Font getMonospace() {
    return INSTANCE.getMonospaceFont();
  }

  private Font getMonospaceFont() {
    if (monospace == null) {
      monospace = new Font(MONOSPACE, getFontSize().getValue());
    }
    return monospace;
  }

  public static Font getBoldArial() {
    return INSTANCE.getBoldArialFont();
  }

  private Font getBoldArialFont() {
    if (boldArial == null) {
      boldArial = Font.font(ARIAL, FontWeight.BOLD, getFontSize().getValue());
    }
    return boldArial;
  }

  public static Font getBoldSerif() {
    return INSTANCE.getBoldSerifFont();
  }

  private Font getBoldSerifFont() {
    if (boldSerif == null) {
      boldSerif = Font.font(SERIF, FontWeight.BOLD, getFontSize().getValue());
    }
    return boldSerif;
  }

  public static Font getPlainArial() {
    return INSTANCE.getPlainArialFont();
  }

  private Font getPlainArialFont() {
    if (plainArial == null) {
      plainArial = new Font(ARIAL, getFontSize().getValue());
    }
    return plainArial;
  }

  public static Font getPlainSerif() {
    return INSTANCE.getPlainSerifFont();
  }

  private Font getPlainSerifFont() {
    if (plainSerif == null) {
      plainSerif = new Font(SERIF, getFontSize().getValue());
    }
    return plainSerif;
  }

  public static Font getBorderTitle() {
    return INSTANCE.getBorderTitleFont();
  }

  private Font getBorderTitleFont() {
    if (borderTitle == null) {
      borderTitle = Font.font(ARIAL, FontPosture.ITALIC, getFontSize().getValue());
    }
    return borderTitle;
  }

  List<MenuBar> mbs = new ArrayList<>();
  List<Node> nodes = new ArrayList<>();
  List<TabPane> tabPanes = new ArrayList<>();
  List<Tab> tabs = new ArrayList<>();

  public static void register(Node control) {
    if (!INSTANCE.nodes.contains(control)) {
      INSTANCE.nodes.add(control);
      if (control instanceof TextInputControl) {
        control.setStyle(INSTANCE.getMonoStyle(INSTANCE.getFontSize().getValue()));
      } else {
        control.setStyle(FontManagerFX.getStyleFor(INSTANCE.getFontSize().getValue()));
      }
    }
  }

  public static void registerTabPane(TabPane tabPane) {
    if (INSTANCE.tabPanes.contains(tabPane)) {
      Logop.debug("Ignoring duplicate FontManagerFX registration of tabPane " + tabPane.getId());
    } else {
      INSTANCE.tabPanes.add(tabPane);
    }
  }

  public static void registerTab(Tab tab) {
    if (INSTANCE.tabs.contains(tab)) {
      Logop.debug("Ignoring duplicate FontManagerFX registration of tab " + tab.getId());
    } else {
      INSTANCE.tabs.add(tab);
    }
  }

  public static void registerMenubar(MenuBar menuBar) {
    if (INSTANCE.mbs.contains(menuBar)) {
      Logop.debug("Ignoring duplicate FontManagerFX registration of menubar " + menuBar.getId());
    } else {
      INSTANCE.mbs.add(menuBar);
    }
  }

  public static void init() {
    setFontSize(INSTANCE.getFontSize());
  }

  public static int getSize() {
    return INSTANCE.getFontSize().getValue();
  }
  
  private FontSize fontSize;

  private FontSize getFontSize() {
    if (fontSize == null) {
      try {
        String savedFontSizeValue = GlobalProperties.getGP(FONT_SIZE_KEY);
        if (savedFontSizeValue == null || savedFontSizeValue.length() < 1) {
          fontSize = FontSize.DEFAULT_SIZE;
          return fontSize;
        }
        fontSize = FontSize.findFromValue(savedFontSizeValue);
      } catch (NumberFormatException ex) {
        Logop.error(ex);
        fontSize = FontSize.DEFAULT_SIZE;
      }
    }
    return fontSize;

  }

  public static void setFontSize(FontSize fontSize) {
    INSTANCE.fontSize = fontSize;
    INSTANCE.updateAllFontSizes();
  }

  private String getMonoStyle(int i) {
    return String.format("-fx-font-size: %dpx; -fx-font-family: 'monospaced';", i);
  }

  private void updateAllFontSizes() {
    String style = getStyleFor(getFontSize().getValue());
    String stylem = getMonoStyle(getFontSize().getValue());
    for (MenuBar mb : mbs) {
      mb.setStyle(style);
    }
    for (Node node : nodes) {
      if (node instanceof TextInputControl) {
        if ("menubarStatusMessage".equals(node.getId())) {
          Platform.runLater(() -> {
            if (node != null) {
              setMenubarLogRecord(logEntry, (TextField) node);
            }
          });
        } else {
          node.setStyle(stylem);
        }
      } else {
        node.setStyle(style);
      }
    }
    for (TabPane tabPane : tabPanes) {
      tabPane.setStyle(style);
    }
    for (Tab tab : tabs) {
      tab.setStyle(style);
    }
    GlobalProperties.putGP(FONT_SIZE_KEY, String.format("%d", getFontSize().getValue()));
    monospace = null;
    boldArial = null;
    plainArial = null;
    boldSerif = null;
    plainSerif = null;
    borderTitle = null;
  }

  public static double getImgRatio() {
    return INSTANCE.getFontSize().getImgRatio();
  }


  public static double getSvgRatio() {
    return INSTANCE.getFontSize().getSvgRatio();
  }

  public static String getCurrentMessageStyle(LogEntry logEntry) {
    String colorStyle = FONT_GREEN;
    if (logEntry.getAlert().equals(Logop.Alert.PURPLE)) {
      colorStyle = FONT_PURPLE;
    } else if (logEntry.getAlert().equals(Logop.Alert.BLUE)) {
      colorStyle = FONT_BLUE;
    } else if (logEntry.getAlert().equals(Logop.Alert.RED)) {
      colorStyle = FONT_RED;
    } else if (logEntry.getAlert().equals(Logop.Alert.ORANGE)) {
      colorStyle = FONT_ORANGE;
    } else if (logEntry.getAlert().equals(Logop.Alert.YELLOW)) {
      colorStyle = FONT_YELLOW;
    } else if (logEntry.getAlert().equals(Logop.Alert.GREEN)) {
      colorStyle = FONT_GREEN;
    }
    return colorStyle + getStyleFor(getSize());
  }

  public static String getStyleFor(int fontSizeValue) {
    return String.format("-fx-font-size: %dpx;", fontSizeValue);
  }

  public static synchronized void setMenubarLogRecord(LogEntry logEntry, TextField menubarStatusMessage) {
    if (logEntry == null) {
      return;
    }
    INSTANCE.logEntry = logEntry;
    if (INSTANCE.logEntry.getMessage() == null) {
      return;
    }
    Platform.runLater(() -> {
      menubarStatusMessage.setStyle(FontManagerFX.getCurrentMessageStyle(INSTANCE.logEntry));
      menubarStatusMessage.setText(INSTANCE.logEntry.getMessage());
    });
  }

}

File: code/shoring/fxapp/src/main/kotlin/com/panopset/fxapp/FontManagerFX.kt
package com.panopset.fxapp

import com.panopset.compat.*
import javafx.application.Platform
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.text.Font
import javafx.scene.text.FontPosture
import javafx.scene.text.FontWeight

object FontManagerFX {
    var fontSize: FontSize = FontSize.DEFAULT_SIZE

    private var monospace: Font = Font(MONOSPACE, fontSize.value.toDouble())
    private var boldArial: Font = Font.font(ARIAL, FontWeight.BOLD, fontSize.value.toDouble())
    private var plainArial: Font = Font(ARIAL, fontSize.value.toDouble())
    private var boldSerif: Font = Font.font(SERIF, FontWeight.BOLD, fontSize.value.toDouble())
    private var plainSerif: Font = Font(SERIF, fontSize.value.toDouble())
    private var borderTitle: Font = Font.font(ARIAL, FontPosture.ITALIC, fontSize.value.toDouble())
//    private var logEntry: LogEntry = LogEntry(AlertColor.GREEN, Level.INFO, "")

    init {
        fontSize = try {
            val savedFontSizeValue = globalPropsGet(FONT_SIZE_KEY).replace("\"", "")
            if (savedFontSizeValue.isEmpty()) {
                FontSize.DEFAULT_SIZE
            }
            FontSize.findFromValue(savedFontSizeValue)
        } catch (ex: NumberFormatException) {
            Logz.errorEx(ex)
            FontSize.DEFAULT_SIZE
        }
    }

    fun getCurrentFontSizeName(): String {
        return fontSize.name
    }

    private fun getMonoStyle(i: Int): String {
        return String.format("-fx-font-size: %dpx; -fx-font-family: 'monospaced';", i)
    }

    fun getCurrentBaseStyle(): String {
        return getStyleFor(fontSize.value)
    }

    fun updateAllFontSizes(fxDoc: FxDoc) {
        val style = getStyleFor(fontSize.value)
        val stylem = getMonoStyle(fontSize.value)
        for (mb in fxDoc.mbs) {
            mb.style = style
        }
        for (node in fxDoc.nodes) {
            if (node is TextInputControl) {
                if (MBSM == node.id) {
                    Platform.runLater {
                        fxDoc.fxDocMessage.refresh()
                    }
                } else {
                    node.style = stylem
                }
            } else {
                node.style = style
            }
        }
        for (tabPane in fxDoc.tabPanes) {
            tabPane.style = style
        }
        for (tab in fxDoc.tabs) {
            tab.style = style
        }
        globalPropsPut( FONT_SIZE_KEY, String.format("%d", fontSize.value))
        monospace = Font(MONOSPACE, fontSize.value.toDouble())
        boldArial = Font.font(ARIAL, FontWeight.BOLD, fontSize.value.toDouble())
        plainArial = Font(ARIAL, fontSize.value.toDouble())
        boldSerif = Font.font(SERIF, FontWeight.BOLD, fontSize.value.toDouble())
        plainSerif = Font(SERIF, fontSize.value.toDouble())
        borderTitle = Font.font(ARIAL, FontPosture.ITALIC, fontSize.value.toDouble())
    }

    fun getMonospace(): Font {
        return monospace
    }

    fun getBoldArial(): Font {
        return boldArial
    }

    fun getBoldSerif(): Font {
        return boldSerif
    }

    fun getPlainArial(): Font {
        return plainArial
    }

    fun getPlainSerif(): Font {
        return plainSerif
    }

    fun getBorderTitle(): Font {
        return borderTitle
    }

    fun register(fxDoc: FxDoc, control: Node) {
        if (!fxDoc.nodes.contains(control)) {
            fxDoc.nodes.add(control)
            if (control is TextInputControl) {
                control.style = getMonoStyle(fontSize.value)
            } else {
                control.style = getStyleFor(fontSize.value)
            }
        }
    }

    fun registerTab(fxDoc: FxDoc, tab: Tab): Tab {
        if (fxDoc.tabs.contains(tab)) {
            Logz.debug("Ignoring duplicate FontManagerFX registration of tab " + tab.id)
        } else {
            fxDoc.tabs.add(tab)
        }
        return tab
    }

    fun registerMenubar(fxDoc: FxDoc, menuBar: MenuBar) {
        if (fxDoc.mbs.contains(menuBar)) {
            Logz.debug("Ignoring duplicate FontManagerFX registration of menubar " + menuBar.id)
        } else {
            fxDoc.mbs.add(menuBar)
        }
    }

    fun getSize(): Int {
        return fontSize.value
    }

    fun setFontSize(fxDoc: FxDoc, fontSize: FontSize) {
        this.fontSize = fontSize
        updateAllFontSizes(fxDoc)
    }

    val imgRatio: Double = fontSize.imgRatio
    val svgRatio: Double = fontSize.svgRatio

//    private fun getCurrentMessageStyle(logEntry: LogEntry): String {
//        var colorStyle = FONT_GREEN
//        if (logEntry.alert == AlertColor.PURPLE) {
//            colorStyle = FONT_PURPLE
//        } else if (logEntry.alert == AlertColor.BLUE) {
//            colorStyle = FONT_BLUE
//        } else if (logEntry.alert == AlertColor.RED) {
//            colorStyle = FONT_RED
//        } else if (logEntry.alert == AlertColor.ORANGE) {
//            colorStyle = FONT_ORANGE
//        } else if (logEntry.alert == AlertColor.YELLOW) {
//            colorStyle = FONT_YELLOW
//        } else if (logEntry.alert == AlertColor.GREEN) {
//            colorStyle = FONT_GREEN
//        }
//        return colorStyle + getStyleFor(size)
//    }

    fun getStyleFor(fontSizeValue: Int): String {
        return String.format("-fx-font-size: %dpx", fontSizeValue)
    }

//    fun setMenubarLogRecord(logEntry: LogEntry, menubarStatusMessage: TextField) {
//        this.logEntry = logEntry
//        Platform.runLater {
//            menubarStatusMessage.style = getCurrentMessageStyle(logEntry)
//        }
//    }
}

const val LUCIDIA = "Lucida Sans Regular"
const val LUCIDIA_BOLD = "Lucida Sans Bold"
const val LUCIDIA_OBLIQUE = "Lucida Sans Oblique"
const val LUCIDIA_BOLD_OB = "Lucida Sans Bold Oblique"
const val L_BRIGHT = "Lucida Bright Regular"
const val L_BRIGHT_BOLD = "Lucida Bright Bold"
const val L_BRIGHT_OBLIQUE = "Lucida Bright Italic"
const val L_BRIGHT_BOLD_OB = "Lucida Bright Bold Italic"
const val MONOSPACE = "Monospace"
const val MONOSPACE_BOLD = "Monospace Bold"
const val MONOSPACE_OBLIQUE = "Lucida Sans Typewriter Oblique"
const val MONOSPACE_BOLD_OB = "Lucida Sans Typewriter Bold Oblique"
const val ARIAL = "Arial"
const val SERIF = "Serif"
const val DIALOG = "Dialog"

//const val FONT_GREEN = "-fx-background-color: #000000; -fx-text-fill: #99ff99;"
//const val FONT_YELLOW = "-fx-background-color: #000000;-fx-text-fill: #ffff66;"
//const val FONT_ORANGE = "-fx-background-color: #000000;-fx-text-fill: #ffaa00;"
//const val FONT_BLUE = "-fx-background-color: #000000;-fx-text-fill: #0000ff;"
//const val FONT_RED = "-fx-background-color: #000000;-fx-text-fill: #ff3333;"
//const val FONT_PURPLE = "-fx-background-color: #000000;-fx-text-fill: #dd33dd;"
const val MBSM = "menubarStatusMessage"
